File: app/src/main/java/cn/woong/opengl/utils/TextResourceReader.java
package cn.woong.opengl.utils;

import android.content.Context;
import android.content.res.Resources;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

public class TextResourceReader {
    public static String readTextFileFromResource(Context context, int resourceId) {
        StringBuilder builder = new StringBuilder();

        try {
            InputStream inputStream = context.getResources().openRawResource(resourceId);
            InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
            BufferedReader bufferedReader = new BufferedReader(inputStreamReader);

            String nextLine;
            while ((nextLine = bufferedReader.readLine()) != null) {
                builder.append(nextLine);
                builder.append('\n');
            }
        } catch (IOException ioException) {
            throw new RuntimeException("Can not open resource: " + resourceId, ioException);
        } catch (Resources.NotFoundException nfe) {
            throw new RuntimeException("Can not find resource: " + resourceId, nfe);
        }

        return builder.toString();
    }
}

File: app/src/main/java/cn/woong/opengl/utils/TextResourceReader.kt
package cn.woong.opengl.utils

import android.content.Context
import android.content.res.Resources

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader

object TextResourceReader {
    /**
     * 从 raw 文件中加载着色器资源
     */
    fun readTextFileFromResource(context: Context, resourceId: Int): String {
        val builder = StringBuilder()

        try {
            val inputStream = context.resources.openRawResource(resourceId)
            val inputStreamReader = InputStreamReader(inputStream)
            val bufferedReader = BufferedReader(inputStreamReader)

            // 逐行读取文件内容
            var nextLine: String? = bufferedReader.readLine()
            while (nextLine != null) {
                builder.append(nextLine)
                builder.append("\n")
                nextLine = bufferedReader.readLine()
            }
        } catch (ioException: IOException) {
            throw RuntimeException("Can not open resource: $resourceId", ioException)
        } catch (nfe: Resources.NotFoundException) {
            throw RuntimeException("Can not find resource: $resourceId", nfe)
        }

        return builder.toString()
    }
}
